From a68a5ee9ec41bc6776dd8decff75c5e729240479 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Cl=C3=A9ment=20Le=20Goffic?= <clement.legoffic@foss.st.com>
Date: Thu, 5 Sep 2024 13:46:42 +0200
Subject: [PATCH 1154/1695] drivers: stm32_iwdg: Add support for extended
 watchdog timeout
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This commit enhances the STM32 IWDG driver to support extended watchdog
timeouts.
It introduces new fields in the `stm32_iwdg_device` structure, updates
the interrupt handler to handle multiple refresh cycles, and improves
the timeout handling logic to manage and calculate extended timeouts
accurately.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 8c6d2a30ac37976314dc3ce2762cf6f6418a449e drivers: stm32_iwdg: Add support for extended watchdog timeout]
Change-Id: I4fc0a02b8f7dba74d86c403bbaec36a421cf24aa
Signed-off-by: Clément Le Goffic <clement.legoffic@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/396400
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Domain-Review: Antonio Maria BORNEO <antonio.borneo@foss.st.com>
Reviewed-by: Antonio Maria BORNEO <antonio.borneo@foss.st.com>
---
 core/drivers/stm32_iwdg.c | 70 +++++++++++++++++++++++++++++++--------
 1 file changed, 57 insertions(+), 13 deletions(-)

diff --git a/core/drivers/stm32_iwdg.c b/core/drivers/stm32_iwdg.c
index 599924246..e4961573c 100644
--- a/core/drivers/stm32_iwdg.c
+++ b/core/drivers/stm32_iwdg.c
@@ -23,6 +23,7 @@
 #include <libfdt.h>
 #include <mm/core_memprot.h>
 #include <sm/sm.h>
+#include <stdint.h>
 #include <stm32_util.h>
 #include <string.h>
 #include <trace.h>
@@ -92,10 +93,13 @@
  *          to reset the system
  * @flags - Property flags for the IWDG instance
  * @timeout - Watchdog elaspure timeout
+ * @saved_nb_int - Saved number of interrupts before panic
+ * @nb_int - Remaining number of interrupts before panic
  * @hw_version - Watchdog HW version
  * @last_refresh - Time of last watchdog refresh
  * @wdt_chip - Wathcdog chip instance
  * @link - Link in registered watchdog instance list
+ * @max_hw_timeout - Maximum hardware timeout
  */
 struct stm32_iwdg_device {
 	struct io_pa_va base;
@@ -107,10 +111,13 @@ struct stm32_iwdg_device {
 	struct rstctrl *reset;
 	uint32_t flags;
 	unsigned long timeout;
+	unsigned long saved_nb_int;
+	unsigned long nb_int;
 	unsigned int hw_version;
 	TEE_Time last_refresh;
 	struct wdt_chip wdt_chip;
 	SLIST_ENTRY(stm32_iwdg_device) link;
+	unsigned long max_hw_timeout;
 };
 
 static SLIST_HEAD(iwdg_dev_list_head, stm32_iwdg_device) iwdg_dev_list =
@@ -144,6 +151,14 @@ static bool iwdg_wdt_is_enabled(struct stm32_iwdg_device *iwdg)
 	return iwdg->flags & IWDG_FLAGS_ENABLED;
 }
 
+static void iwdg_refresh(struct stm32_iwdg_device *iwdg)
+{
+	if (tee_time_get_sys_time(&iwdg->last_refresh))
+		panic();
+
+	io_write32(get_base(iwdg) + IWDG_KR_OFFSET, IWDG_KR_RELOAD_KEY);
+}
+
 /* Return counter value to related to input timeout in seconds, or 0 on error */
 static uint32_t iwdg_timeout_cnt(struct stm32_iwdg_device *iwdg,
 				 unsigned long to_sec)
@@ -206,7 +221,12 @@ static enum itr_return stm32_iwdg_it_handler(struct itr_handler *h)
 	/* Disable early interrupt */
 	stm32_iwdg_it_ack(iwdg);
 
-	panic("Watchdog");
+	if (iwdg->nb_int > 1) {
+		iwdg->nb_int--;
+		io_write32(get_base(iwdg) + IWDG_KR_OFFSET, IWDG_KR_RELOAD_KEY);
+	} else {
+		panic("Watchdog");
+	}
 
 	return ITRR_HANDLED;
 }
@@ -256,14 +276,6 @@ static void iwdg_start(struct stm32_iwdg_device *iwdg)
 	iwdg_wdt_set_enabled(iwdg);
 }
 
-static void iwdg_refresh(struct stm32_iwdg_device *iwdg)
-{
-	if (tee_time_get_sys_time(&iwdg->last_refresh))
-		panic();
-
-	io_write32(get_base(iwdg) + IWDG_KR_OFFSET, IWDG_KR_RELOAD_KEY);
-}
-
 /* Operators for watchdog OP-TEE interface */
 static struct stm32_iwdg_device *wdt_chip_to_iwdg(struct wdt_chip *chip)
 {
@@ -282,7 +294,7 @@ static TEE_Result iwdg_wdt_init(struct wdt_chip *chip,
 
 	/* Be safe and expect any counter to be above 2 */
 	*min_timeout = 3 * IWDG_PRESCALER_256 / rate;
-	*max_timeout = (IWDG_CNT_MASK + 1) * IWDG_PRESCALER_256 / rate;
+	*max_timeout = INT32_MAX;
 
 	return TEE_SUCCESS;
 }
@@ -319,19 +331,48 @@ static void iwdg_wdt_refresh(struct wdt_chip *chip)
 {
 	struct stm32_iwdg_device *iwdg = wdt_chip_to_iwdg(chip);
 
+	iwdg->nb_int = iwdg->saved_nb_int;
 	iwdg_refresh(iwdg);
 }
 
+static void stm32_iwdg_handle_big_timeout(struct stm32_iwdg_device *iwdg,
+					  unsigned long timeout_sec) {
+	unsigned long interval = 0;
+	unsigned long rate = 0;
+	unsigned long n = 0;
+	long w = 0;
+
+	rate = clk_get_rate(iwdg->clk_lsi);
+	iwdg->max_hw_timeout = (IWDG_CNT_MASK + 1) * IWDG_PRESCALER_256 / rate;
+	interval = iwdg->max_hw_timeout - IWDG_ETIMEOUT_SEC;
+
+	if (timeout_sec > IWDG_ETIMEOUT_SEC)
+		n = (timeout_sec - IWDG_ETIMEOUT_SEC) / interval;
+
+	if ((timeout_sec - IWDG_ETIMEOUT_SEC) % interval != 0) {
+		/* Approximate the 'big timeout' */
+		w = ((timeout_sec - IWDG_ETIMEOUT_SEC) / (n + 1)) +
+		    IWDG_ETIMEOUT_SEC;
+		iwdg->timeout = w;
+		n = n + 1;
+	} else {
+		iwdg->timeout = iwdg->max_hw_timeout;
+	}
+
+	iwdg->saved_nb_int = n;
+	iwdg->nb_int = n;
+}
+
 static TEE_Result iwdg_wdt_set_timeout(struct wdt_chip *chip,
 				       unsigned long timeout)
 {
 	struct stm32_iwdg_device *iwdg = wdt_chip_to_iwdg(chip);
 
+	stm32_iwdg_handle_big_timeout(iwdg, timeout);
+
 	if (!iwdg_timeout_cnt(iwdg, timeout))
 		return TEE_ERROR_BAD_PARAMETERS;
 
-	iwdg->timeout = timeout;
-
 	if (iwdg_wdt_is_enabled(iwdg)) {
 		TEE_Result res = TEE_ERROR_GENERIC;
 
@@ -362,7 +403,8 @@ static TEE_Result iwdg_wdt_get_timeleft(struct wdt_chip *chip, bool *is_enabled,
 	if (res)
 		panic();
 
-	time.seconds = iwdg->timeout;
+	time.seconds = (iwdg->timeout - IWDG_ETIMEOUT_SEC) * iwdg->saved_nb_int
+		       + IWDG_ETIMEOUT_SEC;
 	TEE_TIME_ADD(iwdg->last_refresh, time, time);
 	if (TEE_TIME_LE(time, now)) {
 		*timeleft = 0;
@@ -443,6 +485,8 @@ static TEE_Result stm32_iwdg_parse_fdt(struct stm32_iwdg_device *iwdg,
 	if (!iwdg->timeout)
 		return TEE_ERROR_BAD_PARAMETERS;
 
+	stm32_iwdg_handle_big_timeout(iwdg, iwdg->timeout);
+
 	if (!iwdg_timeout_cnt(iwdg, iwdg->timeout)) {
 		EMSG("Timeout %lu not applicable", iwdg->timeout);
 		return TEE_ERROR_BAD_PARAMETERS;
-- 
2.47.2

